### ISSUE TITLE
False positive for accessing enum "instance fields"

### ISSUE THREAD

**Issue by ikonst** [dropbox/mypy-protobuf#65]
For example, let's assume a `Money` message (containing amount and currency) and a `Money.Currency` enum.

This should be valid:
```py
some_money = Money()
some_money.amount = 200
some_money.currency = Money.USD
```
and
```py
some_money.currency = Money.Currency.Value('CAD')
```
However, this should be invalid (but is currently valid):
```py
cad = some_money.currency.Value('CAD')
```

Similarly, this is a false positive:
```py
Money.Currency('CAD')
```
since it finds `int.__init__` although in runtime it would result in:
```
TypeError: 'EnumTypeWrapper' object is not callable
```

One way to solve it would be introduce a `CurrencyValue`:
```py
class Money(google___protobuf___message___Message):
  CurrencyValue = NewType('int')
  class Currency:
    DESCRIPTOR: ...
    ...
  USD = typing___cast(CurrencyValue, 1)
  ...
```
This would be a breaking change since it'll require changing any signature accepting `Money.Currency`. Also, since it exists only in the type stubs, you'd have to quote it, i.e.
```
def do_something(currency: 'Money.CurrencyValue'):
```

Alternatively, there might be some (new?) mypy construct that allows to designate fields and functions as class-access only -- basically the same as its built-in (hardcoded, from what I could tell) support for `enum`.

**Comment by ikonst** [comment]
For a related reason, we cannot declare:
```py
some_currencies: List[Currency] = [Money.USD, Money.CAD]
```
since this would result in:
```
TypeError: Parameters to generic types must be types. Got <google.protobuf.internal.enum_type_wrapper.EnumTypeWrapper object at ...>
```

**Comment by ikonst** [comment]
Another false positive:
```py
Money.Currency('CAD')
```
since it finds `int.__init__` although in runtime it would result in:
```
TypeError: 'EnumTypeWrapper' object is not callable
```

**Comment by ikonst** [comment]
I tried coming up with a solution but ultimately ran into the problem of mypy lacking support for generic metaclasses:

```py
from typing import *

Pb_Enum_Has_No_Init = NewType('Pb_Enum_Has_No_Init', int)
T = TypeVar('T')

class PbEnumMeta(type, Generic[T]):
    # I wish this would work, but ...
    # def __call__(cls, *args, **kwds) -> NoReturn: ...
    
    def Name(cls, number: int) -> str: ...
    def Value(cls, name: str) -> T: ...
    def keys(cls) -> List[str]: ...
    def values(cls) -> List[T]: ...
    def items(cls) -> List[Tuple[str, T]]: ...


class Money:
    # I wish we could use generic metaclasses :(
    class Currency(int, metaclass=PbEnumMeta['Money.Currency']):
        # This hack can work to prevent naive `Money.Currency(1)` invocations
        def __init__(self, value: Pb_Enum_Has_No_Init) -> None: ...

    USD = cast(Currency, 1)
    CAD = cast(Currency, 2)
```

**Comment by nipunn1313** [comment]
I'm not sure I follow. I can try to play with it later to get a better understanding.
Can you provide a full example (with contents of proto file, python callsite, expected mypy error, actual runtime result)

Would be great to have a testcase to really spell it out.

**Comment by nipunn1313** [comment]
After looking into #56, I think I follow this better. I believe #79 should essentially cover this by using `Literal` from mypy.

**Comment by nipunn1313** [comment]
I believe #79 should cover
